docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+
+## Table of Contents
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Table of Contents](#table-of-contents)
+- [Glossary](#glossary)
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Divide and conquer](#divide-and-conquer)
+    - [Tenets](#tenets)
+        - [Same UX](#same-ux)
+        - [Fallback to Immutable rollouts](#fallback-to-immutable-rollouts)
+        - [Clean separation of concern](#clean-separation-of-concern)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+    - [User Stories](#user-stories)
+        - [Story 1](#story-1)
+        - [Story 2](#story-2)
+        - [Story 3](#story-3)
+        - [Story 4](#story-4)
+        - [Story 5](#story-5)
+        - [Story 6](#story-6)
+        - [Story 7](#story-7)
+    - [High level flow](#high-level-flow)
+    - [MachineDeployment updates](#machinedeployment-updates)
+    - [KCP updates](#kcp-updates)
+    - [Machine updates](#machine-updates)
+    - [Infra Machine Template changes](#infra-machine-template-changes)
+    - [Remediation](#remediation)
+    - [API Changes](#api-changes)
+        - [Machine](#machine)
+        - [MachineDeployment](#machinedeployment)
+        - [KCP](#kcp)
+        - [External Update RuntimeExtension](#external-update-runtimeextension)
+            - [`CanUpdateMachine` endpoint](#canupdatemachine-endpoint)
+            - [`UpdateMachine` endpoint](#updatemachine-endpoint)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+- [Additional Details](#additional-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+__In-place Update__: any change to a Machine spec, including the Kubernetes Version, that is performed without deleting the machines and creating a new one.
+
+__External Update Lifecycle Hook__: CAPI Lifecycle Runtime Hook to invoke external update extensions.
+
+__External Update Extension__: Runtime Extension (Implementation) is a component responsible to perform in place updates when  the `External Update Lifecycle Hook` is invoked.
+
+## Summary
+
+The proposal introduces update extensions allowing users to execute custom strategies when performing Cluster API rollouts.
+
+An External Update Extension implementing custom update strategies will report the subset of changes they know how to perform. Cluster API will orchestrate the different extensions, polling the update progress from them.
+
+If the totality of the required changes cannot be covered by the defined extensions, Cluster API will allow to fall back to the current behavior (rolling update).
```

Comment:
The logic can also use fallback scenario in case of timeout or some general condition. It might not scale well with multiple upgraders, but having options here would seem beneficial.

Since the changes are constrained on the single machine, machine replace should still work?

Reply:
> what if the external update tempered with objects in a way that the fallback is no longer possible

You mean there is (or there will be) case that external update can do by rollout update can't? If it happens, we can introduce some verification logic to determine if it can fallback.

Reply:
I'd be in favor of having a possibility to disable fallback to rollout updates. In some cases, users would want only certain fields to be handled in-place, for example, instance tags, if any other fields were changed it should be ok to do rollout update.

Reply:
@neolit123 A couple of clarifications:
* The fallback strategy is not meant for the scenario where the in-place update starts and fails. In this case, the update will remain in "failed" state until either the user manually intervenes or remediation (if configured) kicks in a deletes the failed machine. The fallback strategy is meant for when the external updaters cannot handle the desired update. In other words, when capi detects the need for an update, it queries the external updaters and decides to either start an in-place update or a rolling update (fallback strategy). But once it makes that decision and the update starts, it doesn't switch strategies.
* We were thinking that the fallback strategy would be optional. TBD if opt-in or opt-out, pending the discussion on the API changes.